File: foolish-core-scala/src/main/scala/org/foolish/fvm/scubc/OneShotSearchFiroe.scala
package org.foolish.fvm.scubc

import org.foolish.ast.{AST, SearchOperator}

/**
 * OneShotSearchFiroe performs a one-shot search on a brane (head/tail).
 */
class OneShotSearchFiroe(oneShotSearch: AST.OneShotSearchExpr) extends AbstractSearchFiroe(oneShotSearch, oneShotSearch.operator()) {

  override protected def initialize(): Unit =
    super.initialize()
    val searchExpr = ast.asInstanceOf[AST.OneShotSearchExpr]
    enqueueExprs(searchExpr.anchor())

  override protected def executeSearch(target: BraneFiroe): FIR =
    val targetMemory = target.braneMemory
    if targetMemory.isEmpty then
      return new NKFiroe()

    operator match
      case SearchOperator.HEAD => targetMemory.get(0)
      case SearchOperator.TAIL => targetMemory.getLast
      case _ => throw new IllegalStateException("Unknown one-shot operator: " + operator)

  override def toString: String = ast.toString

  /**
   * Copy constructor for cloneConstanic.
   * Resets search state so the search can be re-executed in a new context.
   */
  private def this(original: OneShotSearchFiroe, newParent: FIR) =
    this(original.ast.asInstanceOf[AST.OneShotSearchExpr])
    setParentFir(newParent)
    // Copy braneMemory contents with cloned items
    original.braneMemory.stream.foreach { fir =>
      val cloned = fir.asInstanceOf[Constanicable].cloneConstanic(this, Some(Nyes.INITIALIZED))
      braneMemory.put(cloned)
      indexLookup.put(cloned, braneMemory.size - 1)
      // For nested FiroeWithBraneMind instances, set up parent memory link
      if cloned.isInstanceOf[FiroeWithBraneMind] then
        cloned.asInstanceOf[FiroeWithBraneMind].ordinated = false
        cloned.asInstanceOf[FiroeWithBraneMind].ordinateToParentBraneMind(this, indexLookup.size - 1)
    }

  override def cloneConstanic(newParent: FIR, targetNyes: Option[Nyes]): FIR =
    if !isConstanic then
      throw IllegalStateException(
        s"cloneConstanic can only be called on CONSTANIC or CONSTANT FIRs, " +
        s"but this FIR is in state: ${getNyes}")
    if isConstant then
      return this  // Share CONSTANT searches
    val copy = new OneShotSearchFiroe(this, newParent)
    targetNyes.foreach(copy.setNyes)
    copy
}

File: foolish-core-java/src/main/java/org/foolish/fvm/ubc/OneShotSearchFiroe.java
package org.foolish.fvm.ubc;

import org.foolish.ast.AST;

/**
 * OneShotSearchFiroe performs a one-shot search on a brane (head/tail).
 */
public class OneShotSearchFiroe extends AbstractSearchFiroe {

    public OneShotSearchFiroe(AST.OneShotSearchExpr oneShotSearch) {
        super(oneShotSearch, oneShotSearch.operator());
    }

    /**
     * Copy constructor for cloneConstanic.
     */
    protected OneShotSearchFiroe(OneShotSearchFiroe original, FIR newParent) {
        super(original, newParent);
    }

    @Override
    protected void initialize() {
        super.initialize();
        AST.OneShotSearchExpr searchExpr = (AST.OneShotSearchExpr) ast;
        storeExprs(searchExpr.anchor());
    }

    @Override
    protected FIR executeSearch(Cursor cursor) {
        ReadOnlyBraneMemory targetMemory = cursor.brane().getBraneMemory();
        if (targetMemory.isEmpty()) {
             return new NKFiroe();
        }

        try {
            return targetMemory.get(cursor.statementIndex());
        } catch (IndexOutOfBoundsException e) {
            return new NKFiroe();
        }
    }

    @Override
    public String toString() {
        return ast.toString();
    }

    @Override
    protected FIR cloneConstanic(FIR newParent, java.util.Optional<Nyes> targetNyes) {
        if (!isConstanic()) {
            throw new IllegalStateException(
                formatErrorMessage("cloneConstanic can only be called on CONSTANIC or CONSTANT FIRs, " +
                    "but this FIR is in state: " + getNyes()));
        }

        if (isConstant()) {
            return this;  // Share CONSTANT searches
        }

        OneShotSearchFiroe copy = new OneShotSearchFiroe(this, newParent);

        if (targetNyes.isPresent()) {
            copy.nyes = targetNyes.get();
        } else {
            copy.nyes = this.nyes;
        }

        return copy;
    }
}
